File: scala/org/softlang/company/Dept.scala
package org.softlang.company

/**
 * A department has a name, a manager and subunits.
 */
class Dept {
  private var name: String = ""
  private var manager: Employee = _
  private var subunits: List[Subunit] = List()

  def getName = name
  def setName(name: String) = {
    this.name = name
  }

  def getManager = manager
  def setManager(manager: Employee) = {
    this.manager = manager
  }

  def getSubunits = subunits
  def setSubunits(subunits: List[Subunit]) = {
    this.subunits = subunits
  }

  def total: Double = getManager.total +
    (if (!subunits.isEmpty) (getSubunits.map(_.total).reduceLeft[Double](_ + _)) else 0)

  def cut: Unit = {
    getManager.cut
    getSubunits.foreach(_.cut)
  }

}

File: struts/src/org/softlang/model/Dept.java
package org.softlang.model;

import java.io.Serializable;
import java.util.LinkedList;
import java.util.List;

/**
 * A department has a name, a manager and a list of subunits
 * 
 */
public class Dept implements Serializable {

	private static final long serialVersionUID = -2008895922177165250L;
	private String name;
	private Employee manager;
	private List<Subunit> subunits;

	public void setSubunits(List<Subunit> subunits) {
		this.subunits = subunits;
	}

	public Dept() {
		name = "";
		subunits = new LinkedList<Subunit>();
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Employee getManager() {
		return manager;
	}

	public void setManager(Employee manager) {
		this.manager = manager;
	}

	public List<Subunit> getSubunits() {
		return subunits;
	}

	public double total() {
		double total = 0;
		total += getManager().total();
		for (Subunit s : getSubunits())
			total += s.total();
		return total;		
	}	
	
	public void cut() {
		getManager().cut();
		for (Subunit s : getSubunits())
			s.cut();
	}	
}
